File: channel/src/skills.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import type { ServerBridge } from "./bridge";

export interface SkillIndexEntry {
  name: string;
  version: string;
  sha256: string;
  size_bytes: number;
  uploaded_by_machine: string | null;
  uploaded_at: string;
  content_type: string;
}

const VERSION_FILE = ".cctui-version";

function skillsRoot(): string {
  return process.env.CCTUI_SKILLS_DIR ?? join(homedir(), ".claude", "skills");
}

function readLocalVersion(root: string, name: string): string | null {
  try {
    return readFileSync(join(root, name, VERSION_FILE), "utf8").trim();
  } catch {
    return null;
  }
}

function writeLocalVersion(root: string, name: string, sha256: string): void {
  try {
    writeFileSync(join(root, name, VERSION_FILE), `${sha256}\n`);
  } catch (err) {
    console.error(`[cctui-channel] write ${VERSION_FILE} for ${name} failed:`, err);
  }
}

/**
 * Fetch `/skills/index` and pull any skill whose server sha256 differs from
 * the local `.cctui-version`. Best-effort: logs and continues on any error.
 */
export async function syncSkills(bridge: ServerBridge): Promise<void> {
  const root = skillsRoot();
  try {
    mkdirSync(root, { recursive: true });
  } catch (err) {
    console.error(`[cctui-channel] mkdir ${root} failed:`, err);
    return;
  }

  let index: SkillIndexEntry[];
  try {
    index = await bridge.getSkillIndex();
  } catch (err) {
    console.error("[cctui-channel] skill index fetch failed:", err);
    return;
  }

  for (const entry of index) {
    const local = readLocalVersion(root, entry.name);
    if (local === entry.sha256) continue;
    try {
      await pullOne(bridge, root, entry);
      writeLocalVersion(root, entry.name, entry.sha256);
      console.error(
        `[cctui-channel] skill synced: ${entry.name} @ ${entry.sha256.slice(0, 12)}`,
      );
    } catch (err) {
      console.error(`[cctui-channel] skill pull failed for ${entry.name}:`, err);
    }
  }
}

async function pullOne(
  bridge: ServerBridge,
  root: string,
  entry: SkillIndexEntry,
): Promise<void> {
  const bytes = await bridge.getSkillBundle(entry.name);
  const tmp = join(
    root,
    `.cctui-skill-${entry.name}-${process.pid}-${Date.now()}.tar.zst`,
  );
  writeFileSync(tmp, bytes);

  // Replace destination atomically: move existing aside, extract, then drop.
  const dest = join(root, entry.name);
  const backup = existsSync(dest)
    ? `${dest}.cctui-old-${process.pid}-${Date.now()}`
    : null;
  try {
    if (backup) {
      await Bun.$`mv ${dest} ${backup}`.quiet();
    }
    const res = await Bun.spawn(["tar", "--zstd", "-C", root, "-xf", tmp], {
      stdout: "pipe",
      stderr: "pipe",
    }).exited;
    if (res !== 0) {
      throw new Error(`tar --zstd -x exited ${res}`);
    }
    if (backup) {
      await Bun.$`rm -rf ${backup}`.quiet();
    }
  } catch (err) {
    // Roll back if possible.
    if (backup && existsSync(backup)) {
      try {
        if (existsSync(dest)) await Bun.$`rm -rf ${dest}`.quiet();
        await Bun.$`mv ${backup} ${dest}`.quiet();
      } catch {}
    }
    throw err;
  } finally {
    try {
      await Bun.$`rm -f ${tmp}`.quiet();
    } catch {}
  }
}

File: crates/cctui-channel/src/skills.rs
//! Skill sync: diff local `.cctui-version` against server index, pull changed
//! bundles, extract with `tar --zstd`. Port of `channel/src/skills.ts`.

use std::path::{Path, PathBuf};
use std::process::Command;

use cctui_proto::api::SkillIndexEntry;

use crate::bridge::Bridge;

const VERSION_FILE: &str = ".cctui-version";

fn skills_root() -> PathBuf {
    if let Ok(dir) = std::env::var("CCTUI_SKILLS_DIR") {
        return PathBuf::from(dir);
    }
    let home = std::env::var_os("HOME").map(PathBuf::from).unwrap_or_else(|| PathBuf::from("."));
    home.join(".claude").join("skills")
}

fn read_local_version(root: &Path, name: &str) -> Option<String> {
    let p = root.join(name).join(VERSION_FILE);
    std::fs::read_to_string(p).ok().map(|s| s.trim().to_string())
}

fn write_local_version(root: &Path, name: &str, sha: &str) {
    let p = root.join(name).join(VERSION_FILE);
    if let Err(err) = std::fs::write(&p, format!("{sha}\n")) {
        tracing::error!(path = %p.display(), error = %err, "write {VERSION_FILE} failed");
    }
}

pub async fn sync(bridge: &Bridge) {
    let root = skills_root();
    if let Err(err) = std::fs::create_dir_all(&root) {
        tracing::error!(path = %root.display(), error = %err, "mkdir skills root failed");
        return;
    }
    let index = match bridge.get_skill_index().await {
        Ok(i) => i,
        Err(err) => {
            tracing::error!(error = %err, "skill index fetch failed");
            return;
        }
    };
    for entry in index {
        let local = read_local_version(&root, &entry.name);
        if local.as_deref() == Some(entry.sha256.as_str()) {
            continue;
        }
        match pull_one(bridge, &root, &entry).await {
            Ok(()) => {
                write_local_version(&root, &entry.name, &entry.sha256);
                tracing::info!(
                    name = %entry.name,
                    sha = %&entry.sha256[..entry.sha256.len().min(12)],
                    "skill synced"
                );
            }
            Err(err) => {
                tracing::error!(name = %entry.name, error = %err, "skill pull failed");
            }
        }
    }
}

async fn pull_one(bridge: &Bridge, root: &Path, entry: &SkillIndexEntry) -> anyhow::Result<()> {
    let bytes = bridge.get_skill_bundle(&entry.name).await?;
    let pid = std::process::id();
    let now = chrono::Utc::now().timestamp_millis();
    let tmp = root.join(format!(".cctui-skill-{}-{pid}-{now}.tar.zst", entry.name));
    std::fs::write(&tmp, &bytes)?;

    let dest = root.join(&entry.name);
    let backup = if dest.exists() {
        Some(root.join(format!("{}.cctui-old-{pid}-{now}", entry.name)))
    } else {
        None
    };

    let result = (|| -> anyhow::Result<()> {
        if let Some(b) = &backup {
            std::fs::rename(&dest, b)?;
        }
        let status =
            Command::new("tar").arg("--zstd").arg("-C").arg(root).arg("-xf").arg(&tmp).status()?;
        if !status.success() {
            anyhow::bail!("tar --zstd -x exited {status}");
        }
        if let Some(b) = &backup {
            let _ = std::fs::remove_dir_all(b);
        }
        Ok(())
    })();

    if let Err(err) = &result {
        if let Some(b) = &backup
            && b.exists()
        {
            let _ = std::fs::remove_dir_all(&dest);
            let _ = std::fs::rename(b, &dest);
        }
        let _ = std::fs::remove_file(&tmp);
        return Err(anyhow::anyhow!("{err}"));
    }

    let _ = std::fs::remove_file(&tmp);
    Ok(())
}
